### DISCUSSION TITLE
What would it take to add GPU support?

### DISCUSSION THREAD

**Discussion by palday** [JuliaStats/MixedModels.jl#789] [Ideas]
> [!NOTE]
> For now, I'll focus on the linear mixed model. While the generalized linear mixed model uses a `LinearMixedModel` internally for storage, re-implementing PIRLS and nAQG on the GPU would require a quite different approach than the direct computation involved in each step of fitting the linear mixed model.

# Background

Broadly speaking, MixedModels.jl's speed on the CPU comes from extensive specialization on matrices with a particular form. Multiple dispatch is very convenient for this: we can specialize the matrix multiplication `A*B` not just for general sparsity but for the particular sparsity pattern of `A` and `B`. This is one of the advancements relative to its predecessor in R, lme4, though both rely on the same basic insight (formulating the model fitting problem as a penalized least squares problem that sparse methods can be applied to instead of the generalized least squares problem that software like `nlme` uses). In particular, we specialize on the sparsity patterns that occur in classical single-membership models: diagonal matrices and block sparse matrices.[^multimembership]

[^multimembership]: Although this specialization enables certain speed ups, it also means that extending support to multimembership models -- such as [lmerMultimember](https://github.com/jvparidon/lmerMultiMember) does for lme4 -- would require rather extensive changes in order to convert things to general sparse matrices. Similarly, the boost from moving to GPU-based methods may be best obtained by swapping to a general sparse representation on the GPU. On the GPU, it may just be faster to multiply everything rather than have the specialized branches that we do on the CPU because GPUs are comparatively bad at branches but fast at matrix multiplication.

## Specialized sparse storage of blocks of `L`

The specialized sparsity patterns are represented using three types:
1. [`Diagonal`](https://docs.julialang.org/en/v1/stdlib/LinearAlgebra/#LinearAlgebra.Diagonal) from the LinearAlgebra stdlib
2. [`BlockedSparse`](https://github.com/JuliaStats/MixedModels.jl/blob/c0b9e2d25cab6e003667f54fb659fcc74b6e4ffe/src/arraytypes.jl#L51-L67)
3. [`UniformBlockDiagonal`](https://github.com/JuliaStats/MixedModels.jl/blob/c0b9e2d25cab6e003667f54fb659fcc74b6e4ffe/src/arraytypes.jl#L1-L8)

Various specializations of `LinearAlgebra.mul!` for these types and their adjoints are defined in [`linalg.jl`](https://github.com/JuliaStats/MixedModels.jl/blob/c0b9e2d25cab6e003667f54fb659fcc74b6e4ffe/src/linalg.jl). Additionally, [`linalg/rankUpdate.jl`](https://github.com/JuliaStats/MixedModels.jl/blob/c0b9e2d25cab6e003667f54fb659fcc74b6e4ffe/src/linalg/rankUpdate.jl) defines `rankupdate!` methods for combinations of these functions. Realistically, we could replace most calls to `rankUpdate!` and thus the associated methods to calls to 5-argument `mul!` and "just" provide a few additional methods of `mul!` for our specialized types. However, `rankUpdate!` predates 5-argument `mul!` by several Julia versions and we never got around to seeing if it would be faster to swap. Finally, [`linalg/logdet.jl`](https://github.com/JuliaStats/MixedModels.jl/blob/c0b9e2d25cab6e003667f54fb659fcc74b6e4ffe/src/linalg/logdet.jl) and [`linalg/cholUnblocked.jl`](https://github.com/JuliaStats/MixedModels.jl/blob/c0b9e2d25cab6e003667f54fb659fcc74b6e4ffe/src/linalg/cholUnblocked.jl) provide specialized implementations of an in-place log determinant and blocked Cholesky factorization, respectively.[^inplace]

[^inplace]: The in-place evaluation makes things very fast (no allocations!), but also creates a few problems for things like automatic differentiation. We use a gradient-free optimizer so this isn't a problem, but it might be interesting to provide a not in-place version for use in automatic differentiation applied to certain quantities of interest used in various derived quantities such as approximations to the denominator degrees of freedom.

All of these types are used to represent blocks of `L`, [the lower Cholesky factor, at the heart of the parameter estimation](https://juliastats.org/MixedModels.jl/stable/optimization/#Linear-Mixed-Effects-Models). Incredibly, the biggest computational expense in each step of the optimization process is the call to [`updateL!`](https://github.com/JuliaStats/MixedModels.jl/blob/c0b9e2d25cab6e003667f54fb659fcc74b6e4ffe/src/linearmixedmodel.jl#L1295-L1334) -- the actual evaluation of the objective function after `updateL!` is extremely fast, even for extremely large and complex models. 

> [!NOTE]
> `L` is currently stored in "fully blocked" form. You can explore the structure of blocks for a given model by examining `BlockDescription(model)`. However, the block structure may be simplified into three blocks:
> - a Diagonal or UniformBlockDiagonal block
> - a dense rectangular block
> - and a dense lower triangular block
> 
> This representation is actually not that much less efficient for some things. For a crossed design with two grouping variables, there is already a lot of fill-in such that the L[2,2] block is already dense (and we know that `updateL!` spends most of its time there[^eul22]. Moreover, it can sometimes be faster to use  dense matrices, even for sparse data, because of vectorization and costs of memory access -- this holds doubly so for GPU. In other words, it maybe worthwhile examining the 3-block-L representation as the representation for use on the GPU. We have [a method for converting the fully blocked representation to the 3-block representation](https://github.com/JuliaStats/MixedModels.jl/blob/c0b9e2d25cab6e003667f54fb659fcc74b6e4ffe/src/linearmixedmodel.jl#L699-L730), but `updateL!` itself still strongly assumes the fully blocked representation.

[^eul22]: https://embraceuncertaintybook.com/largescaleobserved.html#sec-lrgobsmods

## `updateL!`, `ReMat` and `FeMat`

`updateL!` updates `L` using the matrix [`A`](https://github.com/JuliaStats/MixedModels.jl/blob/c0b9e2d25cab6e003667f54fb659fcc74b6e4ffe/src/linearmixedmodel.jl#L15C86-L15C109) (row-major packed lower triangle of `hcat(Z,X,y)'hcat(Z,X,y)`, an efficient way of storing essentially the cross product of the model matrices) and the various [`ReMat`](https://github.com/JuliaStats/MixedModels.jl/blob/c0b9e2d25cab6e003667f54fb659fcc74b6e4ffe/src/remat.jl#L3-L31)s, which are sections of the random-effects model matrix generated by a random-effects term. There is a lot of specialization around the storage and multiplication of these matrices, but my suspicion is that they can't be made too much more optimal by swapping to the GPU. Why? Because their only use in `updateL` is a call to [`copyscaleinflate!`](https://github.com/JuliaStats/MixedModels.jl/blob/c0b9e2d25cab6e003667f54fb659fcc74b6e4ffe/src/remat.jl#L561-L625), which, as far as I know, isn't the slow spot in that function. It would be good to profile calls to `updateL` on moderately complex models to test this theory though.

Finally, one more matrix is used to store the fixed-effects model matrix, [`FeMat`](https://github.com/JuliaStats/MixedModels.jl/blob/c0b9e2d25cab6e003667f54fb659fcc74b6e4ffe/src/Xymat.jl#L89-L114), which is just the fixed-effects model matrix with the response vector (y) appended as an additional column. `FeMat` is a very thin wrapper and its storage supports any `AbstractMatrix`, but we currently only support constructing a classic dense `Matrix` via the formula interface. For some types of very large but very sparse matrices, it might make sense to have an option to make this sparse.

# But how to GPU?

Given all of this, how should we go about adding support for GPU?

First two observations:

- the core part of the fitting process is the call to `updateL!`
- L is currently stored as a vector of blocks, where each block can be any `AbstractMatrix`

Constructing a model that uses GPU storage is then relatively straightforward: we just use GPU matrices instead of CPU matrices in L. However, there are a few issues to overcome there:
- GPU dependencies are very heavyweight, so I would prefer that GPU-specific code be placed into package extensions. This may mean that we need to define additional stubs at the package level that the extensions can then overload appropriately.
- we need to expose this in an opt-in way through the constructor `LinearMixedModel` (and then make sure that `fit` forwards any relevant keyword arguments to the constructor).
- appropriate methods for `rankUpdate!` will need to be defined
- specialized methods for `LinearAlgebra.mul!` may also be relevant]
- the organization into collections of sparse blocks may not be ideal for GPU (neither the number of small blocks nor their individual sparsity)

For the last bit, I strongly suspect that the 3-block-L formulation will be much more convenient for the GPU and may actually be faster in some cases on the CPU. (In particular, I suspect it will be faster on machines with very many cores and a huge amount of memory when dealing with very large models.) So first action item:

> [!TIP]
> Expose an option to swap an entire model to the 3-block-L representation and implement an appropriate method of `updateL` to support this.

Having done that, my next steps would be:

1. provide an opt-in keyword argument to swap to GPU storage, which would always imply 3-block-L storage
2. write appropriate `rankUpdateL!` and `mul!` methods
3. explore when it makes sense to move which blocks to GPU -- maybe only one of the 3 blocks should be on the GPU?

**For the discussion, please take advantage of multiple thread for different questions/comments.**

**Comment by palday** [comment]
One thing I like the approach I proposed here is that it opens us up to better supporting other matrix backends, such as sparse fixed effects or [rectangular full-packed storage](https://github.com/JuliaLinearAlgebra/RectangularFullPacked.jl).

**Reply by ajinkya-k** [reply]
@dmbates and I were chatting about this in person yesterday, and the rectangular packed storage format change came up

**Comment by ajinkya-k** [comment]
@palday I would like to help with this effort

**Comment by ajinkya-k** [comment]
I ran `pprof` on `refit!` and a huge chunk of compute is spent on `dgemm` (which seems to be outside of Julia??), and the largest proportion of time spent in `updateL!` is in `rankUpdate!` and the second largest seems to be `getindex`. Is that what you would have expected?

**Comment by dmbates** [comment]
Yes.  Essentially all the time to fit the model is spent in `updateL!`.  Within `updateL!` the majority of the time is spent in `rankUpdate!` of L[2,2] from L[2,1].  It is actually a downdating operation because of the `-one(T)` in the call to `rankUpdate!` but the 4-argument form of `rankUpdate!` encompasses both update and downdate operations.  (The names come from Linpack, the predecessor to Lapack, subroutines and refer to updating in place the Cholesky factor of X'X when adding a row to X or removing a row from X.  This is a "rank-1" update or downdate.  Here we are doing a "rank-k" downdate.)

For a model fit to a large data set with at least two grouping factors the `rankUpdate!` and `cholUnblocked!` operations on the second row of blocks (`j = 2` in the `updateL!` code) are inevitably the time-consuming steps.  Which of these takes more time depends on how large `q_1` is relative to `q_2`.  If `q_1` is much larger than `q_2` then the `rankUpdate!` step will take longer.

**Comment by dmbates** [comment]
The point I am trying to make is that unless there is a way of using a GPU to accelerate the `rankUpdate!` operation when L[2,1] is sparse there won't be much to be gained from using a GPU.  And this type of operation, at least in its current implementation, is not the sort of thing that GPU's do well.  You need to take each column in L[2,1] as a sparse vector and collect the lower triangle of L[2,2] -= v * v' which involves a lot of manipulation of indices and hopping around the memory locations of L[2,2].  It is not something that is amenable to SIMD-like operations.

**Comment by dmbates** [comment]
To be more specific, if the current approach is to be used the biggest gains will be somehow making the `rankUpdate!` method defined starting at line 67 of `linalg/rankUpdate!` faster.  Even easier it just has to be faster for the case where `C.uplo == 'L'`.

It may be possible to use multiple threads, working on chunks of the columns of the argument `A` but it will be necessary to keep around scratch copies of the argument `C` or set and remove write-locks.  There is a possibility of keeping track of which column of `C` you are currently updating in a thread (the variable `rvj` inside the loop) and locking that which would be more effective than locking individual elements of `C`.

**Comment by ajinkya-k** [comment]
Sorry if this is a basic question but does rank update just perform a rank one update of the L matrix ?

**Comment by dmbates** [comment]
It's a rank-k update.  If you have a vector v then v * v' is a rank-1 matrix.  A rank-2 symmetric matrix is of the form v * u' + u * v' where u and v are non-collinear vectors.  The term "rank-k" just means "more than 2".

**Reply by ajinkya-k** [reply]
got it!

**Comment by dmbates** [comment]
I checked on the insteval model and about 60% of the time spent in `updateL!` is on the call to `rankUpdate!` when `j = 2` and `jj = 1`, because L[2,1] is 2972 by 1128 with 73421 non-zeros.

Using multiple threads for the `rankUpdate!` method with signature
```
rankUpdate!(C::HermOrSym{T,S}, A::SparseMatrixCSC{T}, α, β) where {T,S}
```
would probably be worthwhile.  The idea would be to split the for loop starting with
```
      @inbounds for jj in axes(A, 2)
```
over several threads and for each thread keep track of `rvj` defined within the loop on `enumerate(rangejj)`.  That is the column of `C.data` (also bound to `Cd` in the function) that is being updated.  As long as no other thread is updating that column the threads can proceed.

I'm not sure what would be a good way of keeping track of the locks on columns of C by threads.  It may be possible to use a vector of nthreads integers and store the active column for each thread in those integers.  If the thread sets its active column to zero at the end of the enumerate loop and checks that none of the active columns are `rvj` when it is calculated then it is free to go ahead.  I'm just not sure what the mechanism would be when a collision is detected.  The thread seeking to lock the column has to wait but what causes it to check again?

**Comment by dmbates** [comment]
I realize that I am no longer discussing GPU support so I should probably open up a new issue.  The point is that I don't think there is much to be gained from GPU's if the bottleneck is this updating of a symmetric dense matrix from the columns of a sparse matrix.

**Comment by ajinkya-k** [comment]
Another basic question: When you say `rankUpdate` performs an update of rank `k` is that the dimension of `theta` ?

**Comment by dmbates** [comment]
`k` is just the rank of `L[2,1] * L[2,1]'`, which is at most the number of rows in `L[2,1]`.  If `L[2,1]` had fewer columns than rows (which would not occur in our reordering of the grouping factors) then the rank would be at most the number of columns.

**Reply by ajinkya-k** [reply]
Okay I finally figured this out. Just sat down and stepped through the code and some old fashioned pen and paper work! 

Thanks for the patience
